src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs

```diff
@@ -159,6 +188,316 @@ private string GetRootPathWithGuid_NoLock()
             return new MetadataAsSourceFile(fileInfo.TemporaryFilePath, navigateLocation, documentName, documentTooltip);
         }
 
+        private async Task<Document> DecompileSymbolAsync(Document temporaryDocument, OptionSet options, ISymbol symbol, CancellationToken cancellationToken)
+        {
+            // Get the name of the type the symbol is in
+            var containingOrThis = symbol.GetContainingTypeOrThis();
+            var fullName = GetFullReflectionName(containingOrThis);
+
+            var compilation = await temporaryDocument.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+            // TODO: retrieve path to actual assembly instead of reference assembly
+            var reference = compilation.GetMetadataReference(symbol.ContainingAssembly);
+            // Load the assembly.
+            var ad = AssemblyDefinition.ReadAssembly(reference.Display, new ReaderParameters() { AssemblyResolver = new RoslynAssemblyResolver(compilation) });
+
+            var decompiler = new CSharpDecompiler(ad.MainModule, GetDecompilerSettings(options));
+            var fullTypeName = new FullTypeName(fullName);
+
+            // Try to decompile; if an exception is thrown the caller will handle it
+            var text = decompiler.DecompileTypeAsString(fullTypeName);
+            return temporaryDocument.WithText(SourceText.From(text));
+        }
+
+        private static DecompilerSettings GetDecompilerSettings(OptionSet options)
```

Comment:
Rather than doing this mapping, why not just take the result from ILSpy and just run it through our formatter? Then we don't have to maintain this.

Reply:
➡️ I'm not yet confident that the output of the decompiler will successfully parse. We have a goal to use the Roslyn formatter in the future but I'm not ready to do that quite yet.